### PR TITLE
feat(auth): multiple auth tokens per user

### PR DESCRIPTION
**Pull request by nils-wisiol**
To increase authentication flexibility, this change allows users to
create and name additional authentication tokens. They can be used
independently of the current login session.

This change needs some database migration that IS NOT AUTOMATED.
Automatic migration requires either more priviliges or a bunch of code.
Not worth for the number of deployments (one).

Migrate with e.g.

```
INSERT INTO desecapi_token(`created`, `key`, `user_id`, `name`)
  SELECT `created`, `key`, `user_id`, 'default' FROM authtoken_token;
TRUNCATE TABLE authtoken_token;
```

Failure to apply this will result in all login tokens invalid.

This commit does not break any available API, but only extends it.

### CODE REVIEW

File: api/desecapi/models.py [hunk 1]
```diff
@@ -53,6 +55,9 @@ class Token(rest_framework.authtoken.models.Token):
     )
     name = models.CharField(_("Name"), max_length=64)
 
+    def generate_key(self):
+        return b64encode(urandom(21)).decode('utf-8')
```
Comment (peterthomassen) [inline api/desecapi/models.py]: fyi, utf-8 is default -- no need to remove though

File: api/desecapi/authentication.py [hunk 2]
```diff
@@ -1,9 +1,14 @@
...
 import base64, os
 from rest_framework import exceptions, HTTP_HEADER_ENCODING
-from rest_framework.authtoken.models import Token
 from rest_framework.authentication import BaseAuthentication, get_authorization_header, authenticate
-from desecapi.models import Domain
+from desecapi.models import Domain, Token
+from rest_framework.authentication import TokenAuthentication as RestFrameworkTokenAuthentication
+
+
+
```
Comment (peterthomassen) [inline api/desecapi/authentication.py]: extra newline

File: api/desecapi/authentication.py [hunk 3]
```diff
@@ -96,6 +101,8 @@ def authenticate_credentials(self, userid, key):
 
 
 class IPAuthentication(BaseAuthentication):
+    model = Token
```
Comment (peterthomassen) [inline api/desecapi/authentication.py]: why is this necessary? `IPAuthentication` does nothing with tokens.

File: api/desecapi/models.py [hunk 4]
```diff
@@ -43,6 +44,20 @@ def create_superuser(self, email, password):
...
+    # relation to user is a ForeignKey, so each user can have more than one token
+    user = models.ForeignKey(
+        settings.AUTH_USER_MODEL, related_name='auth_tokens',
+        on_delete=models.CASCADE, verbose_name=_("User")
+    )
+    name = models.CharField(_("Name"), max_length=64)
+
+    class Meta:
+        abstract = False
+        unique_together = (('user', 'name', 'key'),)
```
Comment (peterthomassen) [inline api/desecapi/models.py]: `key` must also be unique, otherwise it may be unclear which user a key points to

File: api/desecapi/models.py [hunk 5]
```diff
@@ -53,6 +55,9 @@ class Token(rest_framework.authtoken.models.Token):
     )
```
Comment (peterthomassen) [inline api/desecapi/models.py]: `key` field `max_length` is too large for what's returned by `generate_key()`. (Adding here because I can only comment on line 53 and after)

File: api/desecapi/models.py [hunk 6]
```diff
@@ -6,8 +6,9 @@
 from desecapi import pdns, mixins
 import datetime, uuid
 from django.core.validators import MinValueValidator
-from rest_framework.authtoken.models import Token
 from collections import OrderedDict
+import rest_framework.authtoken.models
+from django.utils.translation import ugettext_lazy as _
```
Comment (peterthomassen) [inline api/desecapi/models.py]: Why using internationalization? We're not using it anywhere else.

File: api/desecapi/models.py [hunk 7]
```diff
@@ -43,6 +44,20 @@ def create_superuser(self, email, password):
...
 
 
+class Token(rest_framework.authtoken.models.Token):
+    key = models.CharField(_("Key"), max_length=40, db_index=True, unique=True)
+    # relation to user is a ForeignKey, so each user can have more than one token
+    user = models.ForeignKey(
+        settings.AUTH_USER_MODEL, related_name='auth_tokens',
+        on_delete=models.CASCADE, verbose_name=_("User")
+    )
+    name = models.CharField(_("Name"), max_length=64)
```
Comment (peterthomassen) [inline api/desecapi/models.py]: No default? (The migration uses `'default'`.)

File: test/e2e/spec/api_spec.js [hunk 8]
```diff
@@ -881,6 +957,36 @@ describe("API", function () {
...
 
+            describe("tokens/ endpoint", function () {
+
+                var tokenId;
+
+                function createToken () {
+                    return chakram.post('/tokens/', { name: "e2e-token" }).then(function (response) {
+                        expect(response).to.have.status(201);
+                        tokenId = response.body['id'];
+                    })
```
Comment (peterthomassen) [inline test/e2e/spec/api_spec.js]: check that `name` is set correctly

File: api/desecapi/views.py [hunk 9]
```diff
@@ -41,6 +46,52 @@ def get_client_ip(request):
     return request.META.get('REMOTE_ADDR')
 
 
+class TokenCreateView(djoser.views.TokenCreateView):
+
+    def _action(self, serializer):
+        user = serializer.user
+        token = Token(user=user)
```
Comment (peterthomassen) [inline api/desecapi/views.py]: Elsewhere I suggested to have a default token name (because the migration assigns the name `'default'` to existing tokens). Seeing this line now, I agree that a token created upon regular login should have empty name, or name `'login'` or something similar.

File: api/desecapi/views.py [hunk 10]
```diff
@@ -41,6 +46,52 @@ def get_client_ip(request):
     return request.META.get('REMOTE_ADDR')
 
 
+class TokenCreateView(djoser.views.TokenCreateView):
+
+    def _action(self, serializer):
```
Comment (peterthomassen) [inline api/desecapi/views.py]: The original Djoser `login_user()` function in `utils.py` uses `settings.TOKEN_MODEL.objects.get_or_create()`. The right way to handle this would be to assign a non-standard `Manager` to the token model, so that `get_or_create()` does what you want to do (in the new `Token` model). Then, this view (which is essentially code duplication) is unnecessary.

File: api/desecapi/views.py [hunk 11]
```diff
@@ -41,6 +46,52 @@ def get_client_ip(request):
...
+        _, token = auth.TokenAuthentication().authenticate(request)
+        token.delete()
+        user_logged_out.send(
+            sender=request.user.__class__, request=request, user=request.user
+        )
+        return Response(status=status.HTTP_204_NO_CONTENT)
+
+
+class TokenViewSet(mixins.CreateModelMixin,
+                   mixins.DestroyModelMixin,
```
Comment (peterthomassen) [inline api/desecapi/views.py]: I am not sure how exactly this works -- I thought that create and list operations are for list endpoints while the destroy action aims at a specific model instance. Where is that taken care of? I suppose it's some magic interaction with the `SimpleRouter` used in `urls.py`. Can you elaborate briefly?

**Comment by nils-wisiol** [comment]
current docs snapshot: https://a4a.de/.upload/1536244733_index.html

**Comment by nils-wisiol** [comment]
new migration:

```
INSERT INTO desecapi_token(`created`, `key`, `user_id`, `name`. `user_specific_id`)
  SELECT `created`, `key`, `user_id`, 'default', 1 FROM authtoken_token;
TRUNCATE TABLE authtoken_token;
```
